### ISSUE TITLE
Two ideas: replacing sessions and auto-loading tab contents

### ISSUE THREAD

**Issue by elsurudo** [AlexPerathoner/Sessions#5]
Hi,

Just tried out your extension, and I like the simplicity.

What do you think of these ideas:

1. A common workflow for me (and I imagine for others as well) would be to restore a session, change the state (close/open new tabs), and then replace the session so I can continue later. What do you think? Perhaps even an auto-save option for the session. Is this even possible/feasible, though?

2. When I restore a session, the tabs that aren't currently focused have "Untitled" as the name. Is it possible to auto-load the content so that the title loads?

**Comment by AlexPerathoner** [comment]
I'm glad that you like my work C;

Regarding your points:
1. It's a workflow I often follow myself and I've been wanting to implement this feature for a long time but was always too lazy to do it. It's doable and I'll work on it. I'll release a new version in a few days.
I have an idea on how to create the auto-save option... but I'm not sure at all if it works. I'll try to implement this in the future.

2. It's not possible. After Sessions opens all the tabs it's up to Safari to load the content and show the relative title. That said, Safari should replace "Untitled" with the actual title soon after opening the window.
If you close Safari, go to _History>Open all windows of the last session_ does this happen as well?

By the way
When opening issues try to focus on one point only. Having lots of issues open is not a problem, but having multiple aggregate into one might become annoying.
- Alex

**Comment by elsurudo** [comment]
Thanks for the response.

> 1. It's a workflow I often follow myself and I've been wanting to implement this feature for a long time but was always too lazy to do it. It's doable and I'll work on it. I'll release a new version in a few days.
I have an idea on how to create the auto-save option... but I'm not sure at all if it works. I'll try to implement this in the future.

Nice, looking forward to it. Not sure if it's the same idea you had, but looking a bit at the docs of Safari App Extensions, my idea would be to [inject a script](https://developer.apple.com/documentation/safariservices/safari_app_extensions/injecting_a_script_into_a_webpage) into every webpage that would [pass a message to the extension](https://developer.apple.com/documentation/safariservices/safari_app_extensions/passing_messages_between_safari_app_extensions_and_injected_scripts) containing the page data that would then be auto-saved to the session.

But I'm not sure if the functionality is complete enough (needs experimentation for sure), and whether it can also handle the "close" event somehow.

> By the way
When opening issues try to focus on one point only. Having lots of issues open is not a problem, but having multiple aggregate into one might become annoying.

Good point, will keep this in mind in the future.

**Comment by AlexPerathoner** [comment]
Latest release adds the replace feature discussed in point 1.

I'm currently working on the auto-update solution but I'm not convinced. I feel like it's complicating the UI too much... Do you have any idea on how to present this feature?
Still, it is working. I'm currently using a timer which replaces a selected session with the current one but it's not ideal as:
1. If the user is closing tabs one after the other instead of the entire window at once while the timer fires there's a risk that some information get lost;
2. The update is recurrent and therefore does not take place at each event (opening a tab, closing one, changing website). This can also lead to a loss of information. At the same time it also means less energy usage.

**Comment by AlexPerathoner** [comment]
I decided to show a green dot next to the sessions that are being auto-updated.
The auto-update fires every 5 seconds, so no there should be no big loss - I can't imagine someone opening a tab to work and closing the window in the next 5 seconds.
Regarding performance: I tried auto-updating 20 sessions together for a total of about 160 tabs and the cpu usage was always below 2%800 - energy usage was zero. So no problems there.

Here's a beta version. Please try it out and let me know what you think

[Sessions v1.6-beta.app.zip](https://github.com/AlexPerathoner/Sessions/files/5748894/Sessions.v1.6-beta.app.zip)

**Comment by elsurudo** [comment]
Great – trying it out. Thanks for implementing it!

So if I understand correctly, to restore a session in "auto" mode, I need to right-click -> Restore -> Auto-update? Then, I see the green circle, which shows that it's auto-updating.

It would be great if there was a way to toggle this globally, or per-session, but I haven't thought through the best UI for this.

 One potential bug I noticed is that I currently can't rename any of my sessions (not sure if it's related to these changes or not). Basically I right-click on the session and click "Rename", but the UI doesn't change, and there is no cursor that allows me to rename.

**Comment by AlexPerathoner** [comment]
The rename not working was definitely a bug.
I've added a switch like the one already present to globally toggle the auto-update function. I'm not entirely convinced of this design, but at least it works.

I've also fixed the risk of deleting some tabs while closing the window: all tabs closed short before closing the window won't be removed from the session.

[Sessions.app.zip](https://github.com/AlexPerathoner/Sessions/files/5759559/Sessions.app.zip)
